src/TestFramework/XML/SafeDOMXPath.php

```diff
@@ -0,0 +1,149 @@
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+declare(strict_types=1);
+
+namespace Infection\TestFramework\XML;
+
+use DOMDocument;
+use DOMElement;
+use DOMNameSpaceNode;
+use DOMNode;
+use DOMNodeList;
+use DOMXPath;
+use Error;
+use InvalidArgumentException;
+use function sprintf;
+use Webmozart\Assert\Assert;
+
+/**
+ * @internal
+ *
+ * @property DOMDocument $document
+ */
+final readonly class SafeDOMXPath
+{
+    private DOMXPath $xPath;
+
+    public function __construct(
+        public DOMDocument $document,
+    ) {
+        $this->xPath = new DOMXPath($document);
+    }
+
+    public static function fromFile(string $pathname): self
+    {
+        Assert::file($pathname);
+        Assert::readable($pathname);
+
+        $dom = new DOMDocument();
+        $loaded = @$dom->load($pathname);
+
+        Assert::true(
+            $loaded,
+            sprintf(
+                'The file "%s" does not contain valid XML.',
+                $pathname,
+            ),
+        );
+
+        return new self($dom);
+    }
+
+    public static function fromString(string $xml): self
+    {
+        $document = new DOMDocument();
+        $loaded = @$document->loadXML($xml);
+
+        Assert::true(
+            $loaded,
+            sprintf(
+                'The string "%s" is not valid XML.',
+                $xml,
+            ),
+        );
+
+        return new self($document);
+    }
+
+    /**
+     * @return DOMNodeList<DOMNode|DOMNameSpaceNode>
+     */
+    public function queryList(string $query, ?DOMNode $contextNode = null): DOMNodeList
+    {
+        try {
+            $nodes = @$this->xPath->query($query, $contextNode);
+        } catch (Error) {
+            throw new InvalidArgumentException(
+                sprintf(
+                    'The context node passed for the query "%s" is invalid.',
+                    $query,
+                ),
+            );
+        }
+
+        Assert::isInstanceOf(
+            $nodes,
+            DOMNodeList::class,
+            sprintf(
+                'The query "%s" is invalid.',
+                $query,
+            ),
+        );
+
+        return $nodes;
+    }
+
+    public function queryElement(string $query, ?DOMNode $contextNode = null): ?DOMElement
+    {
+        $nodes = $this->queryList($query, $contextNode);
+
+        Assert::true(
+            $nodes->length <= 1,
+            sprintf(
+                'Expected the query "%s" to return a "%s" with no or one node. Got "%s".',
+                $query,
+                DOMNodeList::class,
+                $nodes->length,
+            ),
+        );
+
+        return $nodes[0] ?? null;
+    }
+
+    public function registerNamespace(string $prefix, string $namespace): void
+    {
+        $result = $this->xPath->registerNamespace($prefix, $namespace);
```

Comment:
one extractable improvement: instead of removing a namespace to register a namespace